### DISCUSSION TITLE
Suggestion: a method to find many documents by Id

### DISCUSSION THREAD

**Discussion by meircarlos** [Automattic/mongoose#9947] [Ideas]
I feel that mongoose is missing a basic method that given an array of Ids, returns an array of the same length, in which at each position is the database document or undefined if it doesn't exist.

This is a really basic an interesting method as it allows not only to easily retrieve and know which Ids exist in the database, but also having the values in the same index position of the resulting array can be quite useful for development.

I also believe this method to be necessary as a consequence of many questions that people have made related to this, in which they have been instructed to do the following:
```
const ids = ['4ed3ede8844f0f351100000c', '4ed3f117a844e0471100000d', '4ed3f18132f50c491100000e'];

model.find({
    '_id': { $in: ids }
});
```

When it can be much more easier and useful to do something like:
```
const ids = ['4ed3ede8844f0f351100000c', '4ed3f117a844e0471100000d', '4ed3f18132f50c491100000e'];

model.findManyById(ids);
```

Spoiler alert: what makes this useful is that they would not work in exactly the same way, meaning that the `find` method works more like an SQL query, as the documentation describes, but how I see it the purpose of this method would be to simulate what the `populate` method already does really nicely by populating all the documents of an array of ids. However there is no easy approach for doing this with a simple array of ids.

I know that this could just be implemented as a plugin for all my models, however I'm just stating it here as a suggestion for mongoose and in case anyone else might be interested, my method looks like this:
```
async findManyById(ids) {
    const filter = {
        '_id': {
            $in: ids
        }
    };

    const existingDocuments = await this.find(filter);

    const documents = new Array(ids.length);

    let index;
    for (const document of existingDocuments) {
        index = ids.findIndex(id => document._id.equals(id));

        if (index !== -1) {
            documents[index] = document;
        }
    }

    return documents;
}
```

**Comment by vkarpov15** [comment]
`const docs = await Promise.all(ids.map(id => Model.findById(id)));` should do the trick I think. Only potential issue is that it executes N queries in parallel.

**Reply by meircarlos** [reply]
Yes, that should work, but it just doesn't feel good to make N queries to the db when it can be done with 1?

Maybe it's just me being picky, I don't know.

**Reply by vkarpov15** [reply]
Doing it with 1 query would likely be faster, but if you don't need the extra performance then I wouldn't worry about it. Another alternative would be to do what you did, just use a [Map](https://thecodebarbarian.com/the-80-20-guide-to-maps-in-javascript.html) instead of `findIndex()` to avoid the unnecessary O(n^2) loop

**Reply by meircarlos** [reply]
I probably don't need it, but for some reason it just hurts me to know that it could be done with 1 query haha.

And yeah, sure, using a Map would make it even better, however I will probably implement it with some npm Map package so that that Map keys work with both a String and an ObjectId, hence being able to pass the array of ids as an array made of String or ObjectId without worrying about the types
